### PR TITLE
Deprecate probe; better explain sample/interpolate

### LINKED ISSUE #4571: Clean up and clarify sampling-like filters
### Describe what maintenance you would like added.

There was a discussion on slack on the use of sampling-like filters, i.e. `sample`, `probe`, and `interpolate`.  One issue is that it is hard to figure out when to use which filter.  The other issue is that `probe` has the opposite behavior of `sample` and `interpolate` in regards to order of operation (see below).

### Links to source code.

_No response_

### Pseudocode or Screenshots

```python
import pyvista as pv

small = pv.ImageData(dimensions=(5, 5, 5))
large = pv.ImageData(dimensions=(10, 10, 10))
print(small.n_points)
print(large.n_points)
print(small.probe(large).n_points)  # gives different result
print(small.sample(large).n_points)
print(small.interpolate(large).n_points)
```

This  gives

```txt
125
1000
1000
125
125
```

### PR DESCRIPTION
**Pull request by MatthewFlamm**
### Overview

This PR deprecates probe from the reasons in #4571.  It also gives suggestions on the when to use sample vs interpolate.

### Details

Closed #4571.

TODO:

- [x] Align docstring deprecation with warning msg
- [x] Check that sample supports everything that probe does <- need to add cell locator option
- [x] Add clarification to sample and interpolate docstring
- [x] Add clarification to examples

### CODE REVIEW

**Comment by banesullivan** [comment]
I thought we converged on implementing `eborp()`? 😄

**Comment by MatthewFlamm** [comment]
The only thing that `sample` does not support today that `probe` supports is choosing the Cell Locator.  But, I'm not sure that this is working for probe today anyway.  From `vtkProbeFilter` [docs](https://vtk.org/doc/nightly/html/classvtkProbeFilter.html#a3f8ecf7772d00409715704b2a36b8216), 

> (A prototype is used as an object factory to instantiate an instance of the prototype to perform the FindCell() operation). If a prototype, and a [vtkFindCellStrategy](https://vtk.org/doc/nightly/html/classvtkFindCellStrategy.html) are not defined, the [vtkDataSet::FindCell()](https://vtk.org/doc/nightly/html/classvtkDataSet.html#a2221c10d3c4cca44e82c5ef70e4e1cbd) is used. If a [vtkFindCellStrategy](https://vtk.org/doc/nightly/html/classvtkFindCellStrategy.html) is not defined, then the prototype is used.

Note this part: "If a prototype, and a [vtkFindCellStrategy](https://vtk.org/doc/nightly/html/classvtkFindCellStrategy.html) are not defined,", which in my readings means that you have to set both of these to change the behavior.  Right now we only set the prototype cell locator and not the vtkFindCellStrategy.

If this is right, then I propose not implementing this functionality for `sample` at this time.

**Comment by banesullivan** [comment]
I'm running into a use case where `sample()` cannot reproduce the results that `probe()` is giving:

```py
import pyvista as pv
import numpy as np

points = pv.wrap(pv.Sphere().points)
points['my data'] = np.arange(points.n_points)

surf = points.reconstruct_surface().probe(points)
```

`sample()` fails whereas `probe()` works out of the box. Perhaps this is a parameters issue

**Comment by MatthewFlamm** [comment]
> I'm running into a use case where `sample()` cannot reproduce the results that `probe()` is giving:
> 
> ```python
> import pyvista as pv
> import numpy as np
> 
> points = pv.wrap(pv.Sphere().points)
> points['my data'] = np.arange(points.n_points)
> 
> surf = points.reconstruct_surface().probe(points)
> ```
> 
> `sample()` fails whereas `probe()` works out of the box. Perhaps this is a parameters issue

Something feels off here.  `mesh2.probe(mesh1)` means probe data from mesh2 and put it on points of mesh1.  So you take the data from the reconstructed mesh, which has no data, and try to put it onto the original mesh in this case.  The default of `probe` and `sample` is to continue passing the cell and point data of the source mesh (the mesh which we are putting data onto) to the result.  So you end up with the original data back _without_ a reconstructed surface.  I think you want the reverse, i.e. `points.probe(points.reconstruct_surface())`.  Did you do `surf =p oints.reconstruct_surface().sample(points)` (note that this is not the same as your example since these two filters are swapped) or `surf = points.sample(points.reconstruct_surface())`, which gives me the same as the probe example you gave (although flawed).

These filters both aren't working in the intended because they are cell-wise interpolation, yet you are trying to probe/sample from a point cloud that has no cell topology.  There is a snap to closest cell option that we could expose for the future which would fix this issue for this type of usage..

**Comment by banesullivan** [comment]
Crap, I messed this up. Yes, thanks for all of this work clarifying these APIs because these are all very messy and inconsistent (to the point that I got quite confused)

**Comment by MatthewFlamm** [comment]
This is ready for review.  The only thing I considered but didn't implement yet was to create a new example to contrast the two filters.

Some future looking thoughts for `sample` filter (mostly so I can reference them later):

- Expose `SnapToClosestCell`, which would be useful for the use case above.   Add as an example at that time.
- Wrap as a filter for MultiBlock since vtkResampleWithDataSet supports this.
- Check that we have exposed all the other useful options.

File: pyvista/core/filters/data_set.py [hunk 1]
```diff
@@ -3161,6 +3208,9 @@ def sample(
         if tolerance is not None:
             alg.SetComputeTolerance(False)
             alg.SetTolerance(tolerance)
+        if locator:
+            alg.SetCellLocatorPrototype(locator)
```
Comment (banesullivan) [inline pyvista/core/filters/data_set.py:3212]: I think this is nice to expose this interface for "power users" but should we consider supporting named locators? For example, allowing users to pass the string name of a locator type and automatically generating it for them?

**Review by banesullivan — DISMISSED** [review]
Excellent work adding the clarifying explanations

File: pyvista/core/filters/data_set.py (hunk 1 above)
Comment (MatthewFlamm) [reply pyvista/core/filters/data_set.py:3212]: This was done to make it a like-for-like for existing `probe` filter.  We could in the future make it so the user can add string instead and then we give them the default instantiated locator.  I'll that to the enhancement list in #4658.

File: pyvista/core/filters/data_set.py [hunk 2]
```diff
@@ -3068,11 +3078,21 @@ def probe(
...
         >>> grid = examples.load_uniform()
-        >>> result = grid.probe(mesh)
-        >>> 'Spatial Point Data' in result.point_data
+        >>> result = grid.probe(mesh)  # doctest:+SKIP
+        >>> 'Spatial Point Data' in result.point_data  # doctest:+SKIP
         True
 
         """
+        # deprecated in v0.40.0
+        # remove in v0.43.0
```
Comment (MatthewFlamm) [inline pyvista/core/filters/data_set.py:3087]:
```suggestion
        # deprecated in v0.41.0
        # remove in v0.44.0
```

**Review by tkoyama010 — APPROVED** [review]
LGTM

**Pull request merged**